File: CS/09_Forms/AutomaticField.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Spire.Pdf;
using Spire.Pdf.AutomaticFields;
using Spire.Pdf.Graphics;
using Spire.Pdf.Tables;

namespace AutomaticField
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {

            // Create a new PdfDocument object
            PdfDocument doc = new PdfDocument();

            // Set the author for the document
            doc.DocumentInformation.Author = "Spire.Pdf";

            // Set the margin for the pages
            PdfUnitConvertor unitCvtr = new PdfUnitConvertor();
            PdfMargins margin = new PdfMargins();
            margin.Top = unitCvtr.ConvertUnits(2.54f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Bottom = margin.Top;
            margin.Left = unitCvtr.ConvertUnits(3.17f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Right = margin.Left;

            // Create a section in the document
            PdfSection section = doc.Sections.Add();
            section.PageSettings.Size = PdfPageSize.A4;
            section.PageSettings.Margins = margin;

            // Create a page in the section
            PdfPageBase page = section.Pages.Add();

            // Draw automatic fields on the page
            DrawAutomaticField(page);

            // Save the document to a file named "AutomaticField.pdf"
            doc.SaveToFile("AutomaticField.pdf");
            doc.Close();

            //Launch the Pdf file
            PDFDocumentViewer("AutomaticField.pdf");
        }

        // Method to draw automatic fields on a PDF page
        private void DrawAutomaticField(PdfPageBase page)
        {
            float y = 20;

            // Draw the title for the field list
            PdfBrush brush1 = PdfBrushes.CadetBlue;
            PdfTrueTypeFont font1 = new PdfTrueTypeFont(new Font("Arial", 16f, FontStyle.Bold));
            //////////////////Use the following code for netstandard dlls/////////////////////////
            /*
             PdfTrueTypeFont font1 = new PdfTrueTypeFont("Arial", 16f, PdfFontStyle.Bold, true);
            */
            PdfStringFormat format1 = new PdfStringFormat(PdfTextAlignment.Center);
            page.Canvas.DrawString("Automatic Field List", font1, brush1,
                page.Canvas.ClientSize.Width / 2, y, format1);
            y = y + font1.MeasureString("Automatic Field List", format1).Height;
            y = y + 15;

            // Define an array of field names
            String[] fieldList = new String[]
            {
        "DateTimeField",
        "CreationDateField",
        "DocumentAuthorField",
        "SectionNumberField",
        "SectionPageNumberField",
        "SectionPageCountField",
        "PageNumberField",
        "PageCountField",
        "DestinationPageNumberField",
        "CompositeField"
            };

            PdfTrueTypeFont font = new PdfTrueTypeFont(new Font("Arial", 12f));
            //////////////////Use the following code for netstandard dlls/////////////////////////
            /*
             PdfTrueTypeFont font = new PdfTrueTypeFont("Arial", 12f, PdfFontStyle.Italic, true);
            */
            PdfStringFormat fieldNameFormat = new PdfStringFormat();
            fieldNameFormat.MeasureTrailingSpaces = true;

            // Iterate through each field name in the list
            foreach (String fieldName in fieldList)
            {
                // Draw the field name
                String text = String.Format("{0}: ", fieldName);
                page.Canvas.DrawString(text, font, PdfBrushes.DodgerBlue, 0, y);
                float x = font.MeasureString(text, fieldNameFormat).Width;
                RectangleF bounds = new RectangleF(x, y, 200, font.Height);
                DrawAutomaticField(fieldName, page, bounds);
                y = y + font.Height + 8;
            }
        }

        // Method to draw a specific automatic field on a PDF page
        void DrawAutomaticField(String fieldName, PdfPageBase page, RectangleF bounds)
        {
            PdfTrueTypeFont font = new PdfTrueTypeFont(new Font("Arial", 12f, FontStyle.Italic));
            PdfBrush brush = PdfBrushes.OrangeRed;
            PdfStringFormat format = new PdfStringFormat(PdfTextAlignment.Left, PdfVerticalAlignment.Middle);

            if ("DateTimeField" == fieldName)
            {
                // Create and draw a DateTime field
                PdfDateTimeField field = new PdfDateTimeField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.DateFormatString = "yyyy-MM-dd HH:mm:ss";
                field.Draw(page.Canvas);
            }

            // Repeat the above code block for other field names
            if ("CreationDateField" == fieldName)
            {
                PdfCreationDateField field = new PdfCreationDateField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.DateFormatString = "yyyy-MM-dd HH:mm:ss";
                field.Draw(page.Canvas);
            }

            if ("DocumentAuthorField" == fieldName)
            {
                PdfDocumentAuthorField field = new PdfDocumentAuthorField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.Draw(page.Canvas);
            }


            if ("SectionNumberField" == fieldName)
            {
                PdfSectionNumberField field = new PdfSectionNumberField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.Draw(page.Canvas);
            }

            if ("SectionPageNumberField" == fieldName)
            {
                PdfSectionPageNumberField field = new PdfSectionPageNumberField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.Draw(page.Canvas);
            }

            if ("SectionPageCountField" == fieldName)
            {
                PdfSectionPageCountField field = new PdfSectionPageCountField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.Draw(page.Canvas);
            }

            if ("PageNumberField" == fieldName)
            {
                PdfPageNumberField field = new PdfPageNumberField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.Draw(page.Canvas);
            }

            if ("PageCountField" == fieldName)
            {
                PdfPageCountField field = new PdfPageCountField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.Draw(page.Canvas);
            }

            if ("DestinationPageNumberField" == fieldName)
            {
                PdfDestinationPageNumberField field = new PdfDestinationPageNumberField();
                field.Font = font;
                field.Brush = brush;
                field.StringFormat = format;
                field.Bounds = bounds;
                field.Page = page as PdfNewPage;
                field.Draw(page.Canvas);
            }

            if ("CompositeField" == fieldName)
            {
                PdfSectionPageNumberField field1 = new PdfSectionPageNumberField();
                field1.NumberStyle = PdfNumberStyle.LowerRoman;
                PdfSectionPageCountField field2 = new PdfSectionPageCountField();
                PdfCompositeField fields = new PdfCompositeField();
                fields.Font = font;
                fields.Brush = brush;
                fields.StringFormat = format;
                fields.Bounds = bounds;
                fields.AutomaticFields = new PdfAutomaticField[] { field1, field2 };
                fields.Text = "section page {0} of {1}";
                fields.Draw(page.Canvas);
            }
        }
        private void PDFDocumentViewer(string fileName)
        {
            try
            {
                System.Diagnostics.Process.Start(fileName);
            }
            catch { }
        }

    }
}

File: VB/09_Forms/AutomaticField.vb
Imports Spire.Pdf
Imports Spire.Pdf.AutomaticFields
Imports Spire.Pdf.Graphics

Namespace AutomaticField
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			' Create a new PdfDocument object
			Dim doc As New PdfDocument()

			' Set the author name in the document information
			doc.DocumentInformation.Author = "Spire.Pdf"

			' Create a PdfUnitConvertor object for unit conversion
			Dim unitCvtr As New PdfUnitConvertor()

			' Create a PdfMargins object to set margins
			Dim margin As New PdfMargins()

			' Convert 2.54 centimeters to points and set it as the top margin
			margin.Top = unitCvtr.ConvertUnits(2.54F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)

			' Set the bottom margin equal to the top margin
			margin.Bottom = margin.Top

			' Convert 3.17 centimeters to points and set it as the left margin
			margin.Left = unitCvtr.ConvertUnits(3.17F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)

			' Set the right margin equal to the left margin
			margin.Right = margin.Left

			' Add a section to the document
			Dim section As PdfSection = doc.Sections.Add()

			' Set the page size of the section to A4
			section.PageSettings.Size = PdfPageSize.A4

			' Set the margins of the section
			section.PageSettings.Margins = margin

			' Add a page to the section
			Dim page As PdfPageBase = section.Pages.Add()

			' Call a function to draw automatic fields on the page
			DrawAutomaticField(page)

			' Save the document to a file named "AutomaticField.pdf"
			doc.SaveToFile("AutomaticField.pdf")

			' Close the document
			doc.Close()

			' Launch the Pdf file
			PDFDocumentViewer("AutomaticField.pdf")
		End Sub

		Private Sub DrawAutomaticField(ByVal page As PdfPageBase)
			' Define the initial value for the y-coordinate
			Dim y As Single = 20

			' Create a PdfBrush for the text color
			Dim brush1 As PdfBrush = PdfBrushes.CadetBlue

			' Create a PdfTrueTypeFont object with Arial font, size 16, and bold style
			Dim font1 As New PdfTrueTypeFont(New Font("Arial", 16.0F, FontStyle.Bold))
			' =============================================================================
			' Use the following code for netstandard dlls
			' =============================================================================
			'Dim font1 As New PdfTrueTypeFont("Arial", 16.0F, FontStyle.Bold, True)
			' =============================================================================

			' Create a PdfStringFormat object for text alignment
			Dim format1 As New PdfStringFormat(PdfTextAlignment.Center)

			' Draw the title "Automatic Field List" on the page
			page.Canvas.DrawString("Automatic Field List", font1, brush1, page.Canvas.ClientSize.Width / 2, y, format1)

			' Update the y-coordinate position
			y = y + font1.MeasureString("Automatic Field List", format1).Height
			y = y + 15

			' Define an array of field names
			Dim fieldList() As String = {"DateTimeField", "CreationDateField", "DocumentAuthorField", "SectionNumberField", "SectionPageNumberField", "SectionPageCountField", "PageNumberField", "PageCountField", "DestinationPageNumberField", "CompositeField"}

			' Create a PdfTrueTypeFont object with Arial font and size 12
			Dim font As New PdfTrueTypeFont(New Font("Arial", 12.0F))
			' =============================================================================
			' Use the following code for netstandard dlls
			' =============================================================================
			'Dim font As New PdfTrueTypeFont("Arial", 12.0F)
			' =============================================================================

			' Create a PdfStringFormat object for field name formatting
			Dim fieldNameFormat As New PdfStringFormat()

			' Set the MeasureTrailingSpaces property to true to measure trailing spaces in the field name
			fieldNameFormat.MeasureTrailingSpaces = True

			' Iterate through each field name in the field list
			For Each fieldName As String In fieldList
				' Create the text to display, including the field name followed by a colon
				Dim text As String = String.Format("{0}: ", fieldName)

				' Draw the field name text on the page using the specified font and color
				page.Canvas.DrawString(text, font, PdfBrushes.DodgerBlue, 0, y)

				' Measure the width of the drawn text
				Dim x As Single = font.MeasureString(text, fieldNameFormat).Width

				' Create a rectangle for the bounds of the automatic field
				Dim bounds As New RectangleF(x, y, 200, font.Height)

				' Call a function to draw the specific automatic field within the bounds
				DrawAutomaticField(fieldName, page, bounds)

				' Update the y-coordinate position
				y = y + font.Height + 8
			Next fieldName
		End Sub

		Private Sub DrawAutomaticField(ByVal fieldName As String, ByVal page As PdfPageBase, ByVal bounds As RectangleF)
			' Create a PdfTrueTypeFont object with Arial font, size 12, and italic style
			Dim font As New PdfTrueTypeFont(New Font("Arial", 12.0F, FontStyle.Italic))

			' Create a PdfBrush for the text color
			Dim brush As PdfBrush = PdfBrushes.OrangeRed

			' Create a PdfStringFormat object for text alignment and vertical alignment
			Dim format As New PdfStringFormat(PdfTextAlignment.Left, PdfVerticalAlignment.Middle)

			' Check if the field name is "DateTimeField"
			If "DateTimeField" = fieldName Then
				' Create a new PdfDateTimeField object
				Dim field As New PdfDateTimeField()

				' Set the font of the field to the specified font
				field.Font = font

				' Set the brush of the field to the specified brush
				field.Brush = brush

				' Set the string format of the field to the specified format
				field.StringFormat = format

				' Set the bounds of the field to the specified bounds
				field.Bounds = bounds

				' Set the date format string of the field to "yyyy-MM-dd HH:mm:ss"
				field.DateFormatString = "yyyy-MM-dd HH:mm:ss"

				' Draw the field on the page canvas
				field.Draw(page.Canvas)
			End If

			' Create new PdfCreationDateField object, PdfDocumentAuthorField, PdfSectionNumberField, PdfSectionPageNumberField,
			' PdfSectionPageCountField, PdfPageNumberField£¬PdfPageCountField£¬PdfDestinationPageNumberField£¬CompositeField

			If "CreationDateField" = fieldName Then
				Dim field As New PdfCreationDateField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.DateFormatString = "yyyy-MM-dd HH:mm:ss"
				field.Draw(page.Canvas)
			End If

			If "DocumentAuthorField" = fieldName Then
				Dim field As New PdfDocumentAuthorField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.Draw(page.Canvas)
			End If


			If "SectionNumberField" = fieldName Then
				Dim field As New PdfSectionNumberField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.Draw(page.Canvas)
			End If

			If "SectionPageNumberField" = fieldName Then
				Dim field As New PdfSectionPageNumberField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.Draw(page.Canvas)
			End If

			If "SectionPageCountField" = fieldName Then
				Dim field As New PdfSectionPageCountField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.Draw(page.Canvas)
			End If

			If "PageNumberField" = fieldName Then
				Dim field As New PdfPageNumberField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.Draw(page.Canvas)
			End If

			If "PageCountField" = fieldName Then
				Dim field As New PdfPageCountField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.Draw(page.Canvas)
			End If

			If "DestinationPageNumberField" = fieldName Then
				Dim field As New PdfDestinationPageNumberField()
				field.Font = font
				field.Brush = brush
				field.StringFormat = format
				field.Bounds = bounds
				field.Page = TryCast(page, PdfNewPage)
				field.Draw(page.Canvas)
			End If

			If "CompositeField" = fieldName Then
				Dim field1 As New PdfSectionPageNumberField()
				field1.NumberStyle = PdfNumberStyle.LowerRoman
				Dim field2 As New PdfSectionPageCountField()
				Dim fields As New PdfCompositeField()
				fields.Font = font
				fields.Brush = brush
				fields.StringFormat = format
				fields.Bounds = bounds
				fields.AutomaticFields = New PdfAutomaticField() { field1, field2 }
				fields.Text = "section page {0} of {1}"
				fields.Draw(page.Canvas)
			End If
		End Sub

		Private Sub PDFDocumentViewer(ByVal fileName As String)
			Try
				Process.Start(fileName)
			Catch
			End Try
		End Sub

	End Class
End Namespace
